asapdiscovery-ml/asapdiscovery/ml/schema_v2/trainer.py

```diff
@@ -0,0 +1,514 @@
+                    )
+                else:
+                    run_id_fn.write_text(run_id)
+
+        return run_id
+
+    def initialize(self):
+        """
+        Build the Optimizer and ML Model described by the stored config.
+        """
+
+        # Adjust output_dir and make sure it exists
+        self.output_dir.mkdir(parents=True, exist_ok=True)
+        # Start the W&B process
+        if self.sweep or self.use_wandb:
+            run_id = self.wandb_init()
+            self.output_dir = self.output_dir / run_id
+        self.output_dir.mkdir(exist_ok=True)
+
+        # If sweep or continuing a run, get the optimizer and model config options from
+        #  the W&B config
+        if self.sweep or self.cont:
+            wandb_optimizer_config = wandb.config["optimizer_config"]
+            wandb_model_config = wandb.config["model_config"]
+
+            self.optimizer_config = self.optimizer_config.update(wandb_optimizer_config)
+            self.model_config = self.model_config.update(wandb_model_config)
+
+        # Build the Model
+        self.model = self.model_config.build().to(self.device)
+
+        # Build the Optimizer
+        self.optimizer = self.optimizer_config.build(self.model.parameters())
+
+        # Build early stopping
+        if self.es_config:
+            self.es = self.es_config.build()
+
+        # Build dataset and split
+        self.ds = self.ds_config.build()
+        self.ds_train, self.ds_val, self.ds_test = self.ds_splitter_config.split(
+            self.ds
+        )
+
+        # Set internal tracker to True so we know we can start training
+        self._is_initialized = True
+
+    def train(self) -> (mtenn.model.Model, dict):
+        """
+        Train the model, returning the trained model and loss_dict.
+
+        Returns
+        -------
+        mtenn.model.Model
+            Trained model
+        dict
+            Loss dict
+        """
+        if not self._is_initialized:
+            if self.auto_init:
+                self.initialize()
+            else:
+                raise ValueError("Trainer was not initialized before trying to train.")
+
+        # Save initial model weights for debugging
+        torch.save(self.model.state_dict(), self.output_dir / "init.th")
+
+        # Train for n epochs
+        for epoch_idx in range(self.start_epoch, self.n_epochs):
+            print(f"Epoch {epoch_idx}/{self.n_epochs}", flush=True)
+            if epoch_idx % 10 == 0 and epoch_idx > 0:
+                train_loss = np.mean(
+                    [v["losses"][-1] for v in self.loss_dict["train"].values()]
+                )
+                val_loss = np.mean(
+                    [v["losses"][-1] for v in self.loss_dict["val"].values()]
+                )
+                test_loss = np.mean(
+                    [v["losses"][-1] for v in self.loss_dict["test"].values()]
+                )
+                print(f"Training loss: {train_loss:0.5f}")
```

Comment:
I would also log this if you feel like it.

Reply:
Lets do this as gives a log of loss values in a file  (i know wandb does this, but in a more low tech setting this should work well) something like the following should do the trick. 

```python
from asapdiscovery.data.logging import FileLogger

logger = FileLogger(...).getLogger()
logger.info("loss: XXX")

```

Reply:
I have implemented this now as an option in the `Trainer` schema